### PR TITLE
[NP-681] document dao to replace file/blob daos

### PR DESCRIPTION
**Pull request by KolombetM**
Added Document File
File encoded into base64 String on the client side and put to the documentFileDAO
related pr: https://github.com/nanoPayinc/NANOPAY/pull/9866

### CODE REVIEW

File: src/foam/blob/InputStreamBlob.java [hunk 1]
```diff
@@ -50,6 +50,33 @@ public long read(OutputStream out, long offset, long length) {
...
+
+  public String readAsString(OutputStream out, long offset, long length) {
+    try {
+      if ( offset != pos_ ) {
+        throw new RuntimeException("Offset does not match stream position");
+      }
+      ByteArrayOutputStream data = new ByteArrayOutputStream();
+
+      long read = 0;
+      byte[] buffer = new byte[BUFFER_SIZE];
```
Comment (jlhughes) [inline src/foam/blob/InputStreamBlob.java]: Is there no way to read from a Stream without an artificial hard limit?

File: src/foam/nanos/auth/ProfilePictureView.js [hunk 2]
```diff
@@ -239,19 +239,24 @@ foam.CLASS({
       this.addFile(file);
     },
 
-    function addFile(file) {
+    async function addFile(file) {
       if ( file.size > ( 2 * 1024 * 1024 ) ) {
```
Comment (jlhughes) [inline src/foam/nanos/auth/ProfilePictureView.js]: Flag this for configuration - future, and also multipart file upload.  We would like to handle liquid file ingestion, which are huge files.  Ask Michal how he's uploading those.

File: src/foam/nanos/auth/User.js [hunk 3]
```diff
@@ -309,7 +309,7 @@ foam.CLASS({
       section: 'personal'
     },
     {
-      class: 'foam.nanos.fs.FileProperty',
+      class: 'foam.nanos.fs.DocumentFileProperty',
```
Comment (jlhughes) [inline src/foam/nanos/auth/User.js]: Is there a reason 'File' is still in the name?

**Review by jlhughes — COMMENTED** [review]
Stopped part way through review.
It is possible that when DocumentDAO needs to use the real file system that it uses FileDAO, and otherwise is not concerned with 'files'.   I would think for invoices and other small documents there is no reason to use the file system and we can store these directly in the journal.

File: src/foam/blob/InputStreamBlob.java (hunk 1 above)
Comment (KolombetM) [reply src/foam/blob/InputStreamBlob.java]: Kind of. It can be wrapped with another stream reader which probably will use the same approach inside, because all available read methods that I could find are:
1. read() - Reads a single character
2. read(char[] cbuf, int offset, int length) - Reads characters into a portion of an array.

File: src/foam/nanos/auth/User.js (hunk 3 above)
Comment (KolombetM) [reply src/foam/nanos/auth/User.js]: No, will rename

File: src/foam/nanos/auth/ProfilePictureView.js (hunk 2 above)
Comment (KolombetM) [reply src/foam/nanos/auth/ProfilePictureView.js]: Flag this for configuration - future. How can I flag it?

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 4]
```diff
@@ -17,6 +17,11 @@ foam.CLASS({
     {
       class: 'Class',
       name: 'of'
+    },
+    {
+      class: 'Long',
+      name: 'minBlobSize',
```
Comment (jlhughes) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]: Think this might be clearer if 
maxStringDataSize
Thoughts.

**Review by jlhughes — APPROVED** [review]

File: src/foam/nanos/fs/File.js [hunk 5]
```diff
@@ -69,9 +68,63 @@ foam.CLASS({
         }
         return url;
       }
+    },
+    {
+      class: 'Blob',
+      name: 'data',
+      javaGetter:`
+        if ( dataIsSet_ ) return data_;
+        if ( this.getDataString() != null && this.getDataString() != "" ){
```
Comment (jlhughes) [inline src/foam/nanos/fs/File.js]:
```suggestion
        if ( ! SafetyUtil.isEmpty(this.getDataString()) ) {
```
// foam.util.SafetyUtil.

**Review by jlhughes — COMMENTED** [review]
Looks good.

**Comment by jlhughes** [comment]
@Mykola97 we don't support 'select' on the file dao correct? 
Thinking this is a perfect place for PROJECTION - which can avoid pulling the blob, and we would only retrieve on 'find'.

File: src/foam/nanos/blob/FileService.java [hunk 6]
```diff
@@ -48,7 +49,7 @@ protected void download(X x) {
 
       // find file from file dao
       File file = (File) fileDAO_.find_(x, id);
-      if ( file == null || file.getData() == null ) {
+      if (file == null || file.getData() == null) {
```
Comment (kgrgreer) [inline src/foam/nanos/blob/FileService.java]: Why did you remove these spaces?

File: src/foam/nanos/fs/File.js [hunk 7]
```diff
@@ -69,9 +70,64 @@ foam.CLASS({
...
+          byte[] decodedBytes = Base64.getDecoder().decode(encodedString);
+          InputStream is = new ByteArrayInputStream(decodedBytes);
+          InputStreamBlob blob = new foam.blob.InputStreamBlob(is, decodedBytes.length);
+          return blob;
+        } else {
+          return null;
+        }
+      `,
+      getter: function(){
+        if ( typeof this.dataString != 'undefined' && this.dataString != '' ) {
```
Comment (kgrgreer) [inline src/foam/nanos/fs/File.js]: You can do foam.Undefined.isInstance(this.dataString). That way you don't need to know five different methods in JS to tell if something is the correct type because FOAM has the equivalent method for each type.

File: src/foam/nanos/fs/File.js [hunk 8]
```diff
@@ -69,9 +70,64 @@ foam.CLASS({
...
+          let b64Data = this.dataString.split(',')[1];
+          const b64toBlob = (b64Data, contentType=this.mimeType, sliceSize=512) => {
+            const byteCharacters = atob(b64Data);
+            const byteArrays = [];
+
+            for (let offset = 0; offset < byteCharacters.length; offset += sliceSize) {
```
Comment (kgrgreer) [inline src/foam/nanos/fs/File.js]: spacing doesn't conform to FOAM standard

File: src/foam/nanos/fs/File.js [hunk 9]
```diff
@@ -69,9 +70,64 @@ foam.CLASS({
...
+              const slice = byteCharacters.slice(offset, offset + sliceSize);
+
+              const byteNumbers = new Array(slice.length);
+              for (let i = 0; i < slice.length; i++) {
```
Comment (kgrgreer) [inline src/foam/nanos/fs/File.js]: spacing doesn't conform to FOAM standard

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 10]
```diff
@@ -10,13 +10,21 @@ foam.CLASS({
   extends: 'foam.dao.AbstractDAODecorator',
 
   imports: [
-    'fileDAO'
+    'fileDAO',
```
Comment (kgrgreer) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]: no trailing commas

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 11]
```diff
@@ -10,13 +10,21 @@ foam.CLASS({
...
+  ],
+  requires: [
+    'foam.nanos.fs.File',
   ],
 
   properties: [
```
Comment (kgrgreer) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]: no trailing commas

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 12]
```diff
@@ -27,15 +35,35 @@ foam.CLASS({
...
       var promises = props.map((prop) => {
         var files = prop.f(obj);
-        return Promise.all(files.map((f) => self.fileDAO.put(f)));
+        return Promise.all(files.map(async (f) =>{
+          if ( f.filesize <= this.maxStringDataSize ){
+            f.dataString = await this.encode(f.data.blob);
+            delete f.instance_.data;
+          }
+            a = await self.fileDAO.put(f);
+            debugger;
```
Comment (kgrgreer) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]: remove debugger statement

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 13]
```diff
@@ -27,15 +35,35 @@ foam.CLASS({
...
+          return self.fileDAO.put(f);
+          }
+        ))
       });
 
       return Promise.all(promises).then((values) => {
         props.forEach((prop, i) => {
+        debugger;
```
Comment (kgrgreer) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]: remove

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 14]
```diff
@@ -27,15 +35,35 @@ foam.CLASS({
...
+
+    async function encode(file) {
+      const toBase64 = file => new Promise((resolve, reject) => {
+          const reader = new FileReader();
+          reader.readAsDataURL(file);
+          reader.onload = () => resolve(reader.result);
+          reader.onerror = error => reject(error);
+      });
+      return await toBase64(file);
+    },
```
Comment (kgrgreer) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]: no trailing comma

File: src/foam/u2/view/FileView.js [hunk 15]
```diff
@@ -25,10 +25,11 @@ foam.CLASS({
         .add(this.slot(function(data) {
           return ! data ? this.E('span') : this.E('a')
             .attrs({
-              href: data.data && data.data.blob ?
+              href: (data.data && data.data.blob) || data.dataString ?
                 URL.createObjectURL(data.data.blob) :
```
Comment (kgrgreer) [inline src/foam/u2/view/FileView.js:29]: Couldn't data.data.blob or data.data be null on this line if data.dataString isn't?

File: src/foam/nanos/blob/FileService.java [hunk 16]
```diff
@@ -61,9 +62,13 @@ protected void download(X x) {
...
-      String blobId = ((IdentifiedBlob) file.getData()).getId();
-      blob = getDelegate().find_(x, blobId);
-      long size = blob.getSize();
+      if ( SafetyUtil.isEmpty(file.getDataString()) ) {
+        String blobId = ((IdentifiedBlob) file.getData()).getId();
+        blob = getDelegate().find_(x, blobId);
+      } else {
+        blob = file.getData();
+      }
+        long size = blob.getSize();
```
Comment (kgrgreer) [inline src/foam/nanos/blob/FileService.java]: indentation is wrong

File: src/foam/u2/view/FileView.js (hunk 15 above)
Comment (KolombetM) [reply src/foam/u2/view/FileView.js:29]: They can't be null if data.dataString isn't. I deleted the check

**Review by jlhughes — APPROVED** [review]

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 17]
```diff
@@ -27,7 +35,15 @@ foam.CLASS({
 
       var promises = props.map((prop) => {
         var files = prop.f(obj);
-        return Promise.all(files.map((f) => self.fileDAO.put(f)));
+        return Promise.all(files.map(async (f) =>{
+          if ( f.filesize <= this.maxStringDataSize ){
+            f.dataString = await this.encode(f.data.blob);
+            delete f.instance_.data;
+          }
+            a = await self.fileDAO.put(f);
```
Comment (jlhughes) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]:
```suggestion
          a = await self.fileDAO.put(f);
```

File: src/foam/nanos/fs/File.js [hunk 18]
```diff
@@ -69,9 +70,65 @@ foam.CLASS({
...
+          byte[] decodedBytes = Base64.getDecoder().decode(encodedString);
+          InputStream is = new ByteArrayInputStream(decodedBytes);
+          InputStreamBlob blob = new foam.blob.InputStreamBlob(is, decodedBytes.length);
+          return blob;
+        } else {
+          return null;
+        }
+      `,
+      getter: function(){
+      debugger;
```
Comment (jlhughes) [inline src/foam/nanos/fs/File.js]:
```suggestion
```

File: src/foam/nanos/fs/File.js [hunk 19]
```diff
@@ -69,9 +70,65 @@ foam.CLASS({
...
+          BlobService blobStore = new foam.blob.BlobStore();
+          byte[] decodedBytes = Base64.getDecoder().decode(encodedString);
+          InputStream is = new ByteArrayInputStream(decodedBytes);
+          InputStreamBlob blob = new foam.blob.InputStreamBlob(is, decodedBytes.length);
+          return blob;
+        } else {
+          return null;
+        }
+      `,
+      getter: function(){
```
Comment (jlhughes) [inline src/foam/nanos/fs/File.js]:
```suggestion
      getter: function() {
```

File: src/foam/nanos/fs/File.js [hunk 20]
```diff
@@ -69,9 +70,65 @@ foam.CLASS({
...
+      },
+      /**
+       * When we export this as the CSV, we are trying to create a new object if this property is undefined.
+       * But because this 'Blob' is an interface, we can not instantiate it.
+       *
+       * Provide an adapt function will fix that issue.
+       */
+      adapt: function(oldObj, newObj) {
+        return newObj;
+      },
```
Comment (jlhughes) [inline src/foam/nanos/fs/File.js]:
```suggestion
      }
```

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 21]
```diff
@@ -27,7 +35,15 @@ foam.CLASS({
 
       var promises = props.map((prop) => {
         var files = prop.f(obj);
-        return Promise.all(files.map((f) => self.fileDAO.put(f)));
+        return Promise.all(files.map(async (f) =>{
```
Comment (jlhughes) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]:
```suggestion
        return Promise.all(files.map(async (f) => {
```

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 22]
```diff
@@ -27,7 +35,15 @@ foam.CLASS({
...
+          if ( f.filesize <= this.maxStringDataSize ){
```
Comment (jlhughes) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]:
```suggestion
          if ( f.filesize <= this.maxStringDataSize ) {
```

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 23]
```diff
@@ -27,7 +35,15 @@ foam.CLASS({
...
+            f.dataString = await this.encode(f.data.blob);
+            delete f.instance_.data;
+          }
+            a = await self.fileDAO.put(f);
+          return self.fileDAO.put(f);
+          }
+        ))
```
Comment (jlhughes) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]:
```suggestion
        }))
```

File: src/foam/nanos/fs/FileArrayDAODecorator.js [hunk 24]
```diff
@@ -27,7 +35,15 @@ foam.CLASS({
...
         var files = prop.f(obj);
-        return Promise.all(files.map((f) => self.fileDAO.put(f)));
+        return Promise.all(files.map(async (f) =>{
+          if ( f.filesize <= this.maxStringDataSize ){
+            f.dataString = await this.encode(f.data.blob);
+            delete f.instance_.data;
+          }
+            a = await self.fileDAO.put(f);
+          return self.fileDAO.put(f);
+          }
```
Comment (jlhughes) [inline src/foam/nanos/fs/FileArrayDAODecorator.js]:
```suggestion
```

File: src/foam/nanos/fs/FileDAODecorator.js [hunk 25]
```diff
@@ -35,11 +40,26 @@ foam.CLASS({
 
         if ( ! file ) return a();
 
+        if ( file.size <= this.maxStringDataSize ){
```
Comment (jlhughes) [inline src/foam/nanos/fs/FileDAODecorator.js]:
```suggestion
        if ( file.size <= this.maxStringDataSize ) {
```

File: src/foam/nanos/fs/FileSelectDAO.js [hunk 26]
```diff
@@ -0,0 +1,20 @@
...
+foam.CLASS({
+  package: 'foam.nanos.fs',
+  name: 'FileSelectDAO',
+  extends: 'foam.dao.ProxyDAO',
+
+  methods:[
+    {
+      name: 'select',
+      code: function(x, sink, skip, limit, order, predicate){
+        debugger;
```
Comment (jlhughes) [inline src/foam/nanos/fs/FileSelectDAO.js]: Assume this class is for debugging. Can you remove it from the PR?
